mirgen: lift constant expressions

zerbina opened the pull request:
## Summary

Introduce anonymous constants to the MIR and use them to lift fully
constant expressions into separate constants during AST -> MIR
translation.

This speeds up the compiler, improves efficiency of the generated
code, and also fixes a memory leak affecting the C backend.

## Details

### Anonymous constants

Anonymous constants are constants that don't have a name/interface
attached -- they're data only. `ConstId` represents both user-defined
and anonymous constants, by using the value of the most-significant bit
to distinguish between them: if the bit is set, the lower 31 bit
represent a `DataId`, if it's not set, the `ConstId` is the ID of a
user-defined constant.

`toConstId`, `isAnon`, and `extract` are the procedure for
introspection and creation of anonymous constants. `MirEnv.dataFor`
transparently returns the `DataId` for anonymous constants.

### Constant expression lifting

* constant `set` constructions are always lifted (mirroring what both
  `cgen` and `jsgen` did)
* non-empty, constant-folded `seq` constructions (`nkBracket`) are
  always lifted (mirroring what `cgen` did)
* non-empty constant tuple, object, and array construction are lifted
  only when used in a non-sink context (e.g., as an argument to a non-
  `sink` parameter)
* due to a limitation with `DataTable`, hidden conversions being part
  of an expression make it non-constant for now (achieved with a custom
  `isDeepConstExpr` procedure)

Lifting constant construction appearing in `sink` contexts would alter
semantics, as constants have to copied from (which might result in a
`=copy` call).

Since `mirgen` now does the lifting of `seq` construction, the
workaround of treating MIR `seq` constructions (`mnkConstr` of type
`tySequence`) as constants (because `cgen` lifted them into immutable
C constants) is no longer needed and thus removed.

### Fixes

* with the C backend, using constant-evaluated constructions of
  `seq[T]` where `T` is an aggregate type - `static: @[(1, 2)]`, for
  example - led to a copy and the original `seq` never being freed
  (refer to `tconst_seq_leak`). Removal of the workaround in
  `injectdestructors` fixes this
* `MirEnv.rewind` used the wrong checkpoint for rewinding the `bodies`
  table, which could result in compiler crashes
* `tsplit_into_seq` works now, but not because the underlying issue
  is fixed. Constants `seq` construction not requiring copies only
  means that the problematic `=copy` hook is not used

### `cgen`

* lifting of constant expressions (`exprComplexConst`, `genSetNode`) is
  removed; `mirgen` does this now
* the `dataCache` table is removed. Its most prominent user was the
  expression lifting, and it largely duplicates `DataTable`. The other
  users of `dataCache` use different strategies now:
	* caching default values uses a `SigHash`-based table
	* string values use a dedicated `StrNode` table (using `string`
    directly would require copying the string)
	* `nil` closure literals are expanded to a local C `const`
* anonymous constants are *defined* (not declared) in each C file
  they're used in

### `jsgen`

* lifting of constant `set` construction is removed (it's done by
  `mirgen` now)
* a JavaScript global is emitted for every anonymous constant on use

### `vmgen`

* the code generator and backend already use `DataId` everywhere, so no
  change is necessary
* the `packed_env` packing logic for `nkNilLit` has to work around
  unexpanded `nil` closure literals, by expanding the literal into a
  tuple construction itself. Unexpanded `nil` closure literals didn't
  reach there previously, but now they do

### Performance

The changes have a significant positive impact on:
* compiler speed
* efficiency of the generated code (all backends are affected)

The main reasons for both are that:
* lifting constant expressions in `mirgen` fixes the performance
  regression with field checks introduced by
  `78082ead0e2c5d15a1928359b7dd81dc48925d7a`, where C constants were
  copied into locals, which neither GCC nor MSVC are able to optimize
  away
* constant expression are now also lifted when using the VM and
  JavaScript backends

Compiler speed also improves because of to the reduction in MIR code,
as less MIR code means:
* less nodes that MIR passes need to scan/inspect/analyze/etc.
* less nodes that the code generators need to generate code for

Benchmarking the compiler compiling the compiler at
`3eb7c736384de0b62c88c0dcedbb910fce47f190`  (with 
`hyperfine --warump 1 "nim c --compileOnly --verbosity:0 --hints:off --warnings:off compiler/nim.nim"`
)
yields:
```
before: 14.54 s ± 0.14  (100%)
now:    13.65 s ± 0.05  ( 94%)
```

zerbina committed:
mirtrees: support anonymous constants

`ConstId` is changed to represent both user-defined and anonymous
constants. The most-significant bit indicates which of the two a
`ConstId` represents: if the bit is set, the lower 31 bits represent
a `DataId`, otherwise the `ConstId` identifies a user-defined
constant.

This encoding is backwards-compatible (>= 2^31 user-defined constants
is highly unlikely if not impossible).

zerbina committed:
mirgen: lift constant expressions

This is essentially the same as what `cgen` currently does, but with
the difference that `mirgen` can - and does - disable the lifting when
an expression is used in a sink context.

Not lifting an expression used in a sink context is preferred, as
constants have to be copied, whereas the raw construction can
construct the value in place (no copy).

As a temporary workaround for `DataTable` not support hidden
conversions in constant expressions, a custom implementation of
`isDeepConstExpr` is used.

zerbina committed:
mirgen: lift constant sets in `in` expressions

The special code generation for `in` calls is changed such that
constant sets reach `genx`, where they're lifted into anonymous
constants.

zerbina committed:
mirenv: `dataFor` supports anonymous constants

The `DataId` is transparently extracted from the `ConstId`.

zerbina committed:
backends: use `scanExpr`

Replace all usage of `discoverFromValueAst` with `scanExpr`.

zerbina committed:
cgen: implement support for anonymous constants

Same as with the old `dataCache` mechanism, the resulting C constants
are **not** shared across modules.

zerbina committed:
cgen: remove the constant expression lifting

It's all done in `mirgen` now.

zerbina committed:
cgen: remove `dataCache` and associated routines

zerbina committed:
cgen: don't use `dataCache` for string literals

String literals use a dedicated table instead, removing another user
of `dataCache`.

zerbina committed:
cgen: don't use `dataCache` for nil closures

`nil` closure values - outside of constant expression - are now
assigned to a function-local constant, instead of the module-level
data cached being used for this. It's simpler, and also removes one
user of `dataCache`.

zerbina committed:
cgen: remove the last usage of `rawConstExpr`

Use a type-based caching mechanism for default value expressions,
removing the last user of `rawConstExpr`. The `rawConstExpr` procedure
is removed.

This also removes the last user of `dataCache`, which will be removed
next.

zerbina committed:
jsgen: implement support for anonymous constant

The first use of an anonymous constant results in the emission of a
JavaScript global storing the data. All usages of the anonymous
constant are then translated to a usage of said global.

zerbina committed:
utils: pretty-printer support for anon constants

Anonymous constants are always rendered as a raw `DataId`.

zerbina committed:
jsgen: remove constant `set` handling

Constant set constructions are already lifted into constants by
`mirgen`.

zerbina committed:
mir: remove seq-constructor special-casing

One less hack. Since `cgen` no longer lifts sequence constructions into
constants, they no longer have to conservatively be treated as non-
owning.

zerbina committed:
tests: add a  `static seq` leak regression test

Lifting constant expression in `mirgen` and the subsequent removal of
the `seq`-constructor special-casing fixes static `seq` values with
aggregate elements resulting in memory leaks when using the C backend.

zerbina committed:
packed_env: work around unexpected AST

Since arbitrary constant expressions are now turned into constant data,
some unexpanded `nil` literals representing closures reach the
`packed_env` serialization.

Closures are expected to be serialized into a tuple construction
(`(Closure (NilLit) (NilLit))`, in the `nil` case), so the
serialization logic manually does the translation for now.

In the future, "expanding" the literal has to happen earlier.

zerbina committed:
tests: update the `--expandArc`-using tests

zerbina committed:
fix: incorrect rewinding for `bodies`

The `bodies` sequence was rewound using the `data` checkpoint, but this
is wrong! It synchronized with `constants`, so it needs to use the
`const` checkpoint.

saem reviewed (approved):
This was really easy to read, partly I'm getting more familiar with the MIR, but there are plenty of broad simplifications happening.

zerbina committed:
Merge branch 'devel' into mirgen-lift-constant-expressions

zerbina committed:
tests: enable `tsplit_into_seq`

The underlying issue is *not* fixed, but with constant `seq`
constructions in sink contexts no longer requiring copying, the
automatic `=copy` hook that triggers the code generation error is not
used anymore.

zerbina commented:
The `tsplit_into_seq` test was unexpectedly successful.

Investigating why this is the case showed that it's a side-effect of `var x = @[]` not expanding to a call of the seq's `=copy` hook. In the case of the `tsplit_into_seq` test, the lifted copy hook was responsible for the code generation failure, meaning that the test works now (even though the code generator limitation underlying the previous failure still exists).

zerbina commented:
/merge
